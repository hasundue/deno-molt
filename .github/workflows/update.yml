name: Update

on:
  schedule:
    - cron: '0 0 * * *'
  workflow_dispatch:

permissions:
  contents: write
  pull-requests: write

jobs:
  update:
    name: Update
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        uses: actions/checkout@v4

      - name: Configure Git
        run: |
          git config --global user.name '${{ github.actor }}'
          git config --global user.email '${{ github.actor }}@users.noreply.github.com'

      - name: Run Molt
        run: deno task update --commit --pre-commit test --summary title.txt --report body.txt

      - name: Push changes
        run: git push

      - name: Create Pull Request
        run: |
          gh pr create --title "$(cat title.txt)" --body "$(cat body.txt)"
