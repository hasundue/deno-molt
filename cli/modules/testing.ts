import {
  CommandStub,
  FileSystemFake,
  LatestVersionStub,
  ReadTextFileStub,
  WriteTextFileStub,
} from "@molt/lib/testing";

/**
 * Enables all test stubs.
 */
export default function () {
  LatestVersionStub.create({
    "deno.land/std": "0.218.2",
    "deno_graph": "0.69.7",
    "node-emoji": "2.1.3",
    "@luca/flag": "1.0.1",
    "@std/": "0.218.2",
  });
  const fs = new FileSystemFake();
  ReadTextFileStub.create(fs, { readThrough: true });
  WriteTextFileStub.create(fs);
  Deno.Command = CommandStub.create("git");
}
