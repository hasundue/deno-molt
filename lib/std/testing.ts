export {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  it,
} from "https://deno.land/std@0.208.0/testing/bdd.ts";
export {
  assertSpyCall,
  assertSpyCallArg,
  assertSpyCalls,
  type ConstructorSpy,
  type ExpectedSpyCall,
  type Spy,
  spy,
  type SpyCall,
  type Stub,
  stub,
} from "https://deno.land/std@0.208.0/testing/mock.ts";
export {
  createAssertSnapshot,
} from "https://deno.land/std@0.208.0/testing/snapshot.ts";
