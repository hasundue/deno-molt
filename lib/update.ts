import { distinct } from "./std/collections.ts";
import { fromFileUrl } from "./std/path.ts";
import {
  createGraph,
  type CreateGraphOptions,
  init as initDenoGraph,
  load as defaultLoad,
  type ModuleJson,
} from "./x/deno_graph.ts";
import { findFileUp, toPath, toUrl } from "./path.ts";
import {
  ImportMap,
  ImportMapResolveResult,
  readImportMapJson,
  tryReadFromJson,
} from "./import_map.ts";
import {
  type Dependency,
  parse,
  resolveLatestVersion,
  toUrl as dependencyToUrl,
  type UpdatedDependency,
} from "./dependency.ts";

type DependencyJson = NonNullable<ModuleJson["dependencies"]>[number];

type If<T extends boolean, A, B> = T extends true ? A : B;

/**
 * Representation of an update to a dependency.
 */
export interface DependencyUpdate<IsMapped extends boolean = boolean> {
  /** Properties of the dependency being updated. */
  from: Dependency;
  /** Properties of the updated dependency. */
  to: UpdatedDependency;
  /**
   * The code of the dependency. Note that `type` in the DependencyJSON
   * is merged into `code` here for convenience.
   */
  code: {
    /** The original specifier of the dependency appeared in the code. */
    specifier: string;
    span: If<IsMapped, undefined, NonNullable<DependencyJson["code"]>["span"]>;
  };
  /** The full path to the module that imports the dependency.
   * @example "/path/to/mod.ts" */
  referrer: string;
  /** Information about the import map used to resolve the dependency. */
  map: If<
    IsMapped,
    {
      /** The full path to the import map used to resolve the dependency.
       * @example "/path/to/import_map.json" */
      source: string;
    } & ImportMapResolveResult<true>,
    undefined
  >;
}

class DenoGraph {
  static #initialized = false;

  static async ensureInit() {
    if (this.#initialized) {
      return;
    }
    await initDenoGraph();
    this.#initialized = true;
  }
}

export interface CollectOptions {
  /**
   * The working directory to resolve relative paths.
   * If not specified, the current working directory is used.
   * At present, this option is only used to find the import map.
   *
   * @example "/path/to/project"
   */
  cwd?: string | URL;
  /**
   * The path to the import map used to resolve dependencies.
   * If not specified, molt will automatically find deno.json or deno.jsonc
   * in the current working directory or parent directories.
   *
   * @example
   * ```ts
   * const updates = await DependencyUpdate.collect("mod.ts", {
   *   importMap: "import_map.json"
   *   // -> Use import_map.json in the current directory
   * });
   * ```
   */
  importMap?: string | URL;
  /**
   * A function to filter out dependencies.
   *
   * @example
   * ```ts
   * const updates = await DependencyUpdate.collect("mod.ts", {
   *   ignore: (dep) => dep.name === "deno.land/std"
   *   // -> Ignore all dependencies from deno.land/std
   * });
   * ```
   */
  ignore?: (dependency: Dependency) => boolean;
  /**
   * A function to pick dependencies.
   *
   * @example
   * ```ts
   * const updates = await DependencyUpdate.collect("mod.ts", {
   *   only: (dep) => dep.name === "deno.land/std"
   *   // -> Only pick dependencies from deno.land/std
   * });
   * ```
   */
  only?: (dependency: Dependency) => boolean;
}

/**
 * Collect dependencies from the given module(s) or Deno configuration file(s).
 * Local submodules are also checked recursively.
 *
 * @param from - The path(s) to the file(s) to collect dependencies from.
 * @param options - Options to customize the behavior.
 * @returns The list of dependencies.
 *
 * @example
 * ```ts
 * collect("mod.ts")
 * // -> Collect dependencies from mod.ts and its local submodules.
 * ```
 * @example
 * ```ts
 * collect("deno.json")
 * // -> Collect dependencies from the import map specified in deno.json
 * ```
 */
export async function collect(
  from: string | URL | (string | URL)[],
  options: CollectOptions = {},
): Promise<DependencyUpdate[]> {
  const froms = [from].flat();
  const urls = froms.map((path) => toUrl(path));

  const importMapPath = options.importMap ??
    await findFileUp(options.cwd ?? Deno.cwd(), "deno.json", "deno.jsonc");

  const importMap = importMapPath
    ? await tryReadFromJson(toUrl(importMapPath))
    : undefined;

  await DenoGraph.ensureInit();
  const graph = await createGraph(urls, {
    load,
    resolve: importMap?.resolveInner,
  });

  const updates: DependencyUpdate[] = [];
  await Promise.all([
    ...graph.modules
      .filter((m) => m.kind === "esm")
      .flatMap((m) =>
        m.dependencies?.map(async (dependency) => {
          const update = await _createDependencyUpdate(
            dependency,
            m.specifier,
            { ...options, importMap },
          );
          return update ? updates.push(update) : undefined;
        })
      ),
    ...graph.modules
      .filter((m) => m.kind === "asserted" && m.mediaType === "Json")
      .map(async (m) => {
        const results = await _collectFromImportMap(m.specifier, options);
        updates.push(...results);
      }),
  ]);
  return updates.sort((a, b) => a.to.name.localeCompare(b.to.name));
}

const load: NonNullable<CreateGraphOptions["load"]> = async (
  specifier,
) => {
  const url = new URL(specifier); // should not throw
  switch (url.protocol) {
    case "node:":
    case "npm:":
    case "jsr:":
      return {
        kind: "external",
        specifier,
      };
    case "http:":
    case "https:":
      return {
        kind: "external",
        specifier,
      };
    case "file:":
      return await defaultLoad(specifier);
    default:
      throw new Error(`Unsupported protocol: ${url.protocol}`);
  }
};

/**
 * Create a DependencyUpdate from the given dependency.
 * @param dependencyJson - The dependency to create an update from.
 * @param referrer - The URL of the module that imports the dependency.
 * @param options - Options to customize the behavior.
 * @returns The created DependencyUpdate.
 */
async function _createDependencyUpdate(
  dependencyJson: DependencyJson,
  referrer: string,
  options?: Pick<CollectOptions, "ignore" | "only"> & {
    importMap?: ImportMap;
  },
): Promise<DependencyUpdate | undefined> {
  const specifier = dependencyJson.code?.specifier ??
    dependencyJson.type?.specifier;
  if (!specifier) {
    throw new Error(
      `The dependency ${dependencyJson.specifier} in ${
        fromFileUrl(referrer)
      } has no resolved specifier.`,
      { cause: dependencyJson },
    );
  }
  const mapped = options?.importMap?.resolve(
    dependencyJson.specifier,
    referrer,
  ) as ImportMapResolveResult<true> | undefined;
  const dependency = parse(new URL(mapped?.value ?? specifier));
  if (options?.ignore?.(dependency)) {
    return;
  }
  if (options?.only && !options.only(dependency)) {
    return;
  }
  const latest = await resolveLatestVersion(dependency);
  if (!latest || latest.version === dependency.version) {
    return;
  }
  const span = dependencyJson.code?.span ?? dependencyJson.type?.span;
  if (!span) {
    throw new Error(
      `The dependency ${dependencyJson.specifier} in ${
        fromFileUrl(referrer)
      } has no span.`,
    );
  }
  return {
    from: dependency,
    to: latest,
    code: {
      // We prefer to put the original specifier here.
      specifier: dependencyJson.specifier,
      span,
    },
    referrer: toPath(referrer),
    map: mapped
      ? {
        source: options!.importMap!.path,
        ...mapped,
      }
      : undefined,
  };
}

async function _collectFromImportMap(
  specifier: ModuleJson["specifier"],
  options: Pick<CollectOptions, "ignore" | "only">,
): Promise<DependencyUpdate[]> {
  const json = await readImportMapJson(new URL(specifier));
  const updates: DependencyUpdate[] = [];
  await Promise.all(
    Object.entries(json.imports).map(
      async ([key, value]): Promise<DependencyUpdate | undefined> => {
        if (!URL.canParse(value)) {
          return;
        }
        const dependency = parse(new URL(value));
        if (options.ignore?.(dependency)) {
          return;
        }
        if (options.only && !options.only(dependency)) {
          return;
        }
        const latest = await resolveLatestVersion(dependency);
        if (!latest || latest.version === dependency.version) {
          return;
        }
        updates.push({
          from: dependency,
          to: latest,
          code: {
            specifier: value,
            span: undefined,
          },
          referrer: toPath(specifier),
          map: {
            source: toPath(specifier),
            resolved: value,
            key,
            value: dependencyToUrl(latest),
          },
        });
      },
    ),
  );
  return updates;
}

export type VersionChange = {
  from?: string;
  to: string;
};

export function getVersionChange(
  dependencies: DependencyUpdate[],
): VersionChange | undefined {
  const modules = distinct(dependencies.map((d) => d.to.name));
  if (modules.length > 1) {
    // Cannot provide a well-defined version prop
    return;
  }
  const tos = distinct(dependencies.map((d) => d.to.version));
  if (tos.length > 1) {
    throw new Error(
      "Multiple target versions are specified for a single module",
    );
  }
  const froms = distinct(dependencies.map((d) => d.from.version));
  return {
    from: froms.length === 1 ? froms[0] : undefined,
    to: tos[0],
  };
}
