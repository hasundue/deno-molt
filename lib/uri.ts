import { isAbsolute, relative, resolve, toFileUrl } from "./std/path.ts";
import { Brand } from "./types.ts";

export type DefaultProtocol<Scheme extends string> = Scheme extends
  "http" | "https" ? `${Scheme}://`
  : Scheme extends "file" ? `${Scheme}:///`
  : `${Scheme}:`;

export type URI<
  Scheme extends string = string,
  Protocol extends string = DefaultProtocol<Scheme>,
> = `${Protocol}${string}`;

export type RelativePath = Brand<string, "RelativePath">;
export type AbsolutePath = Brand<string, "AbsolutePath">;

export const URI = {
  from(path: string): URI<"file"> {
    let url: URL;
    try {
      url = new URL(path);
    } catch {
      return toFileUrl(
        isAbsolute(path) ? path : resolve(path),
      ).href as URI<"file">;
    }
    if (url.protocol !== "file:") {
      throw new TypeError();
    }
    return url.href as URI<"file">;
  },
  relative(uri: URI<"file">): RelativePath {
    return relative(Deno.cwd(), new URL(uri).pathname) as RelativePath;
  },
  absolute(uri: URI<"file">): AbsolutePath {
    return resolve(new URL(uri).pathname) as AbsolutePath;
  },
  ensure<S extends string>(...schemes: S[]): (uri: string) => URI<S> {
    return (uri) => {
      let url: URL;
      try {
        url = new URL(uri);
      } catch {
        throw new TypeError(`Invalid URI: ${uri}`);
      }
      if (schemes.includes(url.protocol.split(":")[0] as S)) {
        return uri as URI<S>;
      }
      throw new TypeError(`Unexpected URI protocol: ${url.protocol}`);
    };
  },
  is<S extends string>(uri: string, scheme: S): uri is URI<S> {
    try {
      const url = new URL(uri);
      return url.protocol.split(":")[0] === scheme;
    } catch {
      return false;
    }
  },
};
