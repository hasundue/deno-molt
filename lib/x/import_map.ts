export {
  ImportMap,
  type ImportMapJson,
  parseFromJson,
} from "https://deno.land/x/import_map@v0.19.1/mod.ts";
