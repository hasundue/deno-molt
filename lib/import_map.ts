import { maxBy } from "./std/collections.ts";
import { parse as parseJsonc } from "./std/jsonc.ts";
import { type ImportMapJson, parseFromJson } from "./x/import_map.ts";
import { is } from "./x/unknownutil.ts";
import type { Maybe } from "./types.ts";
import { URI } from "./uri.ts";
import { URIScheme } from "./types.ts";

export type { ImportMapJson };

export interface ImportMapResolveResult {
  /** The full specifier resolved from the import map. */
  specifier: URI<URIScheme>;
  from?: string;
  to?: string;
}

export interface ImportMap {
  // TODO: Accept a remote URL
  specifier: URI<"file">;
  resolve(specifier: string, referrer: string): Maybe<ImportMapResolveResult>;
  resolveSimple(specifier: string, referrer: string): string;
}

export const ImportMap = {
  readFromJson,
};

const isImportMapJson = is.ObjectOf({
  imports: is.RecordOf(is.String),
});

const isImportMapReferrer = is.ObjectOf({
  importMap: is.String,
});

// This implementation is ridiculously inefficient, but we prefer not to reimplement the whole
// import_map module. Maybe we should rathre patch rust code of the import_map module.
async function readFromJson(url: URL): Promise<Maybe<ImportMap>> {
  const data = await Deno.readTextFile(url.pathname);
  if (data.length === 0) return;
  const json = parseJsonc(data);
  if (isImportMapReferrer(json)) {
    // The json seems to be deno.json or deno.jsonc referencing an import map.
    return readFromJson(new URL(json.importMap, url));
  }
  if (!isImportMapJson(json)) {
    // The json does not include an import map.
    return undefined;
  }
  const inner = await parseFromJson(url, json);
  return {
    specifier: URI.from(url.href),
    resolve(specifier, referrer) {
      const resolved = inner.resolve(specifier, referrer);
      if (resolved === specifier) {
        // The specifier is not resolved by the import map.
        return undefined;
      }
      // Find which key is used for the resolution.
      const replacement = maxBy(
        Object.entries(json.imports)
          .map(([from, to]) => ({ from, to }))
          .filter(({ to }) => resolved.includes(to)),
        ({ to }) => to.length,
      );
      if (!replacement) {
        // The specifier should be a file path
        URI.ensure("file")(resolved);
      }
      return {
        specifier: URI.ensure(...URIScheme.values)(resolved),
        ...replacement,
      };
    },
    resolveSimple: inner.resolve.bind(inner),
  };
}
