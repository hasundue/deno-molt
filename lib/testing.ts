import {
  assertSpyCall,
  assertSpyCallArg,
  ConstructorSpy,
  ExpectedSpyCall,
  Spy,
  spy,
  Stub,
  stub,
} from "./std/testing.ts";
import { AssertionError } from "./std/assert.ts";
import { URI } from "./uri.ts";

export function createCommandStub(): ConstructorSpy<
  Deno.Command,
  ConstructorParameters<typeof Deno.Command>
> {
  const CommandSpy = spy(Deno.Command);
  return class extends CommandSpy {
    #output: Deno.CommandOutput = {
      code: 0,
      stdout: new Uint8Array(),
      stderr: new Uint8Array(),
      success: true,
      signal: null,
    };
    outputSync() {
      return this.#output;
    }
    output() {
      return Promise.resolve(this.#output);
    }
    spawn() {
      return new Deno.ChildProcess();
    }
    static clear() {
      this.calls = [];
    }
  };
}

export class FileSystemFake extends Map<URI<"file">, string> {}

export const ReadTextFileStub = {
  create(
    fs: FileSystemFake,
    options?: {
      readThrough?: boolean;
    },
  ): Stub {
    const original = Deno.readTextFile;
    return stub(
      Deno,
      "readTextFile",
      async (path) => {
        return fs.get(URI.from(path)) ?? options?.readThrough
          ? await original(path)
          : _throw(new Deno.errors.NotFound(`File not found: ${path}`));
      },
    );
  },
};
export type ReadTextFileStub = ReturnType<typeof ReadTextFileStub.create>;

export const WriteTextFileStub = {
  create(
    fs: FileSystemFake,
  ) {
    return stub(
      Deno,
      "writeTextFile",
      // deno-lint-ignore require-await
      async (path, data) => {
        fs.set(URI.from(path), data.toString());
      },
    );
  },
};
export type WriteTextFileStub = ReturnType<typeof WriteTextFileStub.create>;

/** Asserts that a spy is called as expected at any index. */
export function assertFindSpyCall<
  Self,
  Args extends unknown[],
  Return,
>(
  spy: Spy<Self, Args, Return>,
  expected: ExpectedSpyCall<Self, Args, Return>,
) {
  const call = spy.calls.find((_, index) => {
    try {
      assertSpyCall(spy, index, expected);
      return true;
    } catch {
      return false;
    }
  });
  if (!call) {
    throw new AssertionError("Expected spy call does not exist");
  }
  return call;
}

export function assertFindSpyCallArg<
  Self,
  Args extends unknown[],
  Return,
  ExpectedArg,
>(
  spy: Spy<Self, Args, Return>,
  argIndex: number,
  expected: ExpectedArg,
) {
  const call = spy.calls.find((_, index) => {
    try {
      assertSpyCallArg(spy, index, argIndex, expected);
      return true;
    } catch {
      return false;
    }
  });
  if (!call) {
    throw new AssertionError("Expected spy call does not exist");
  }
  return call;
}

/** Utility function to throw an error. */
function _throw(error: Error): never {
  throw error;
}
