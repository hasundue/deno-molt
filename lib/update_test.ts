// deno-lint-ignore-file no-explicit-any
import { beforeAll, describe, it } from "./std/testing.ts";
import {
  assertEquals,
  assertExists,
  assertNotEquals,
  assertObjectMatch,
  assertThrows,
} from "./std/assert.ts";
import { URI } from "./uri.ts";
import { _create, createVersionProp, DependencyUpdate } from "./update.ts";
import { ImportMap } from "./import_map.ts";

describe("_create", () => {
  it("https://deno.land/std", async () => {
    const update = await _create({
      specifier: "https://deno.land/std@0.1.0/version.ts",
      code: {
        specifier: "https://deno.land/std@0.1.0/version.ts",
      } as any,
    }, URI.from("test/fixtures/direct-import/mod.ts"));
    assertExists(update);
  });
  it("https://deno.land/std - no semver", async () => {
    const update = await _create({
      specifier: "https://deno.land/std/version.ts",
      code: {
        specifier: "https://deno.land/std/version.ts",
      } as any,
    }, URI.from("test/fixtures/direct-import/mod.ts"));
    assertEquals(update, undefined);
  });
  it("https://deno.land/x/deno_graph", async () => {
    const update = await _create({
      specifier: "https://deno.land/x/deno_graph@0.1.0/mod.ts",
      code: {
        specifier: "https://deno.land/x/deno_graph@0.1.0/mod.ts",
      } as any,
    }, URI.from("test/fixtures/direct-import/mod.ts"));
    assertExists(update);
  });
  it("npm:node-emoji", async () => {
    const update = await _create({
      specifier: "npm:node-emoji@1.0.0",
      code: {
        specifier: "npm:node-emoji@1.0.0",
      } as any,
    }, URI.from("test/fixtures/direct-import/mod.ts"));
    assertExists(update);
  });
});

describe("_create - with import map", () => {
  let importMap: ImportMap;
  beforeAll(async () => {
    importMap = (await ImportMap.readFromJson(
      URI.from("../test/fixtures/import-map/deno.json", import.meta.url),
    ))!;
  });
  it("std/version.ts", async () => {
    const update = await _create(
      {
        specifier: "std/version.ts",
        code: {
          specifier: "https://deno.land/std@0.200.0/version.ts",
        } as any,
      },
      URI.from("test/fixtures/import-map/mod.ts"),
      { importMap },
    );
    assertExists(update);
    assertObjectMatch(update, {
      name: "deno.land/std",
      version: {
        from: "0.200.0",
        // to: "0.203.0",
      },
      path: "/version.ts",
      specifier: "https://deno.land/std@0.200.0/version.ts",
      code: { specifier: "std/version.ts" },
      referrer: URI.from("test/fixtures/import-map/mod.ts"),
      map: {
        source: URI.from("test/fixtures/import-map/deno.json"),
        from: "std/",
        to: "https://deno.land/std@0.200.0/",
      },
    });
  });
});

describe("collect", () => {
  it("direct import", async () => {
    const updates = await DependencyUpdate.collect(
      "./test/fixtures/direct-import/mod.ts",
    );
    assertEquals(updates.length, 4);
  });
  it("import map", async () => {
    const updates = await DependencyUpdate.collect(
      "./test/fixtures/import-map/mod.ts",
      {
        importMap: "./test/fixtures/import-map/deno.json",
      },
    );
    assertEquals(updates.length, 4);
  });
});

describe("applyToModule", () => {
  let updates: DependencyUpdate[];
  let content: string;
  beforeAll(async () => {
    updates = await DependencyUpdate.collect(
      "./test/fixtures/direct-import/mod.ts",
    );
    content = await Deno.readTextFile("./test/fixtures/direct-import/mod.ts");
  });
  it("https://deno.land/x/deno_graph", () => {
    const update = updates.find((update) =>
      update.specifier.includes("deno.land/x/deno_graph")
    )!;
    const result = DependencyUpdate.applyToModule(
      update,
      content,
    );
    assertExists(result);
    assertNotEquals(result, content);
  });
  it("npm:node-emoji", () => {
    const update = updates.find((update) =>
      update.specifier.includes("node-emoji")
    )!;
    const result = DependencyUpdate.applyToModule(
      update,
      content,
    );
    assertExists(result);
    assertNotEquals(result, content);
  });
});

describe("applyToImportMap", () => {
  let updates: DependencyUpdate[];
  let content: string;
  beforeAll(async () => {
    updates = await DependencyUpdate.collect(
      "./test/fixtures/import-map/mod.ts",
      { importMap: "test/fixtures/import-map/deno.json" },
    );
    content = await Deno.readTextFile("test/fixtures/import-map/deno.json");
  });
  it("deno_graph", () => {
    const update = updates.find((update) =>
      update.code?.specifier === "deno_graph"
    )!;
    const result = DependencyUpdate.applyToImportMap(
      update,
      content,
    );
    assertExists(result);
    assertNotEquals(result, content);
  });
});

describe("createVersionProps()", () => {
  it("single version", () => {
    assertEquals(
      createVersionProp([
        {
          name: "deno_graph",
          version: { from: "0.0.1", to: "0.1.0" },
        },
      ] as any),
      {
        from: "0.0.1",
        to: "0.1.0",
      },
    );
  });
  it("multiple versions with different names", () => {
    assertEquals(
      createVersionProp([{
        name: "deno_graph",
        version: { from: "0.0.1", to: "0.1.0" },
      }, {
        name: "node-emoji",
        version: { from: "0.0.1", to: "0.1.0" },
      }] as any),
      undefined,
    );
  });
  it("multiple versions with different `from`s and a common `to`", () => {
    assertEquals(
      createVersionProp([
        {
          name: "deno_graph",
          version: { from: "0.0.1", to: "0.1.0" },
        },
        {
          name: "deno_graph",
          version: { from: "0.0.2", to: "0.1.0" },
        },
      ] as any),
      {
        from: undefined,
        to: "0.1.0",
      },
    );
  });
  it("multiple versions with a common `from` and `to`", () => {
    assertEquals(
      createVersionProp([
        {
          name: "deno_graph",
          version: { from: "0.0.1", to: "0.2.0" },
        },
        {
          name: "deno_graph",
          version: { from: "0.0.1", to: "0.2.0" },
        },
      ] as any),
      {
        from: "0.0.1",
        to: "0.2.0",
      },
    );
  });
  it("multiple versions with a common `from` and different `to`s", () => {
    assertThrows(() =>
      createVersionProp([
        {
          name: "deno_graph",
          version: { from: "0.0.1", to: "0.1.0" },
        },
        {
          name: "deno_graph",
          version: { from: "0.0.1", to: "0.2.0" },
        },
      ] as any)
    );
  });
});
