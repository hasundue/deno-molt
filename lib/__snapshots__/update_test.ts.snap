export const snapshot = {};

snapshot[`_create > https://deno.land/std 1`] = `
{
  code: {
    span: undefined,
    specifier: "https://deno.land/std@0.1.0/version.ts",
  },
  name: "deno.land/std",
  path: "/version.ts",
  specifier: {
    from: "https://deno.land/std@0.1.0/version.ts",
    to: "https://deno.land/std@0.204.0/version.ts",
  },
  version: {
    from: "0.1.0",
    to: "0.204.0",
  },
}
`;

snapshot[`_create > https://deno.land/std - unversioned 1`] = `
{
  code: {
    span: undefined,
    specifier: "https://deno.land/std/version.ts",
  },
  name: "deno.land/std",
  path: "/version.ts",
  specifier: {
    from: "https://deno.land/std/version.ts",
    to: "https://deno.land/std@0.204.0/version.ts",
  },
  version: {
    from: undefined,
    to: "0.204.0",
  },
}
`;

snapshot[`_create > https://deno.land/x/deno_graph 1`] = `
{
  code: {
    span: undefined,
    specifier: "https://deno.land/x/deno_graph@0.1.0/mod.ts",
  },
  name: "deno.land/x/deno_graph",
  path: "/mod.ts",
  specifier: {
    from: "https://deno.land/x/deno_graph@0.1.0/mod.ts",
    to: "https://deno.land/x/deno_graph@0.59.1/mod.ts",
  },
  version: {
    from: "0.1.0",
    to: "0.59.1",
  },
}
`;

snapshot[`_create > npm:node-emoji 1`] = `
{
  code: {
    span: undefined,
    specifier: "npm:node-emoji@1.0.0",
  },
  name: "node-emoji",
  path: "",
  specifier: {
    from: "npm:node-emoji@1.0.0",
    to: "npm:node-emoji@2.1.0",
  },
  version: {
    from: "1.0.0",
    to: "2.1.0",
  },
}
`;

snapshot[`_create > npm:node-emoji - unversioned 1`] = `
{
  code: {
    span: undefined,
    specifier: "npm:node-emoji",
  },
  name: "node-emoji",
  path: "",
  specifier: {
    from: "npm:node-emoji",
    to: "npm:node-emoji@2.1.0",
  },
  version: {
    from: undefined,
    to: "2.1.0",
  },
}
`;

snapshot[`collect > direct import 1`] = `
{
  code: {
    span: {
      end: {
        character: 66,
        line: 0,
      },
      start: {
        character: 24,
        line: 0,
      },
    },
    specifier: "https://deno.land/std@0.200.0/version.ts",
  },
  name: "deno.land/std",
  path: "/version.ts",
  specifier: {
    from: "https://deno.land/std@0.200.0/version.ts",
    to: "https://deno.land/std@0.204.0/version.ts",
  },
  version: {
    from: "0.200.0",
    to: "0.204.0",
  },
}
`;

snapshot[`collect > direct import 2`] = `
{
  code: {
    span: {
      end: {
        character: 74,
        line: 1,
      },
      start: {
        character: 28,
        line: 1,
      },
    },
    specifier: "https://deno.land/x/deno_graph@0.50.0/mod.ts",
  },
  name: "deno.land/x/deno_graph",
  path: "/mod.ts",
  specifier: {
    from: "https://deno.land/x/deno_graph@0.50.0/mod.ts",
    to: "https://deno.land/x/deno_graph@0.59.1/mod.ts",
  },
  version: {
    from: "0.50.0",
    to: "0.59.1",
  },
}
`;

snapshot[`collect > direct import 3`] = `
{
  code: {
    span: {
      end: {
        character: 46,
        line: 2,
      },
      start: {
        character: 24,
        line: 2,
      },
    },
    specifier: "npm:node-emoji@1.0.0",
  },
  name: "node-emoji",
  path: "",
  specifier: {
    from: "npm:node-emoji@1.0.0",
    to: "npm:node-emoji@2.1.0",
  },
  version: {
    from: "1.0.0",
    to: "2.1.0",
  },
}
`;

snapshot[`collect > direct import 4`] = `
{
  code: {
    span: {
      end: {
        character: 66,
        line: 0,
      },
      start: {
        character: 24,
        line: 0,
      },
    },
    specifier: "https://deno.land/std@0.200.0/version.ts",
  },
  name: "deno.land/std",
  path: "/version.ts",
  specifier: {
    from: "https://deno.land/std@0.200.0/version.ts",
    to: "https://deno.land/std@0.204.0/version.ts",
  },
  version: {
    from: "0.200.0",
    to: "0.204.0",
  },
}
`;

snapshot[`collect > import map 1`] = `
{
  code: {
    span: {
      end: {
        character: 66,
        line: 0,
      },
      start: {
        character: 24,
        line: 0,
      },
    },
    specifier: "https://deno.land/std@0.200.0/version.ts",
  },
  name: "deno.land/std",
  path: "/version.ts",
  specifier: {
    from: "https://deno.land/std@0.200.0/version.ts",
    to: "https://deno.land/std@0.204.0/version.ts",
  },
  version: {
    from: "0.200.0",
    to: "0.204.0",
  },
}
`;

snapshot[`collect > import map 2`] = `
{
  code: {
    span: {
      end: {
        character: 40,
        line: 1,
      },
      start: {
        character: 28,
        line: 1,
      },
    },
    specifier: "deno_graph",
  },
  name: "deno.land/x/deno_graph",
  path: "/mod.ts",
  specifier: {
    from: "https://deno.land/x/deno_graph@0.50.0/mod.ts",
    to: "https://deno.land/x/deno_graph@0.59.1/mod.ts",
  },
  version: {
    from: "0.50.0",
    to: "0.59.1",
  },
}
`;

snapshot[`collect > import map 3`] = `
{
  code: {
    span: {
      end: {
        character: 36,
        line: 2,
      },
      start: {
        character: 24,
        line: 2,
      },
    },
    specifier: "node-emoji",
  },
  name: "node-emoji",
  path: "",
  specifier: {
    from: "npm:node-emoji@1.0.0",
    to: "npm:node-emoji@2.1.0",
  },
  version: {
    from: "1.0.0",
    to: "2.1.0",
  },
}
`;

snapshot[`collect > import map 4`] = `
{
  code: {
    span: {
      end: {
        character: 40,
        line: 0,
      },
      start: {
        character: 24,
        line: 0,
      },
    },
    specifier: "std/version.ts",
  },
  name: "deno.land/std",
  path: "/version.ts",
  specifier: {
    from: "https://deno.land/std@0.200.0/version.ts",
    to: "https://deno.land/std@0.204.0/version.ts",
  },
  version: {
    from: "0.200.0",
    to: "0.204.0",
  },
}
`;

snapshot[`applyToModule > https://deno.land/x/deno_graph 1`] = `
'import { VERSION } from "https://deno.land/std@0.200.0/version.ts";
import { createGraph } from "https://deno.land/x/deno_graph@0.59.1/mod.ts";
import { emojify } from "npm:node-emoji@1.0.0";
import { noop } from "./lib.ts";
'
`;

snapshot[`applyToModule > npm:node-emoji 1`] = `
'import { VERSION } from "https://deno.land/std@0.200.0/version.ts";
import { createGraph } from "https://deno.land/x/deno_graph@0.50.0/mod.ts";
import { emojify } from "npm:node-emoji@2.1.0";
import { noop } from "./lib.ts";
'
`;

snapshot[`applyToImportMap > deno_graph 1`] = `
'{
  "imports": {
    "std/": "https://deno.land/std@0.200.0/",
    "deno_graph": "https://deno.land/x/deno_graph@0.59.1/mod.ts",
    "node-emoji": "npm:node-emoji@1.0.0",
    "/": "./"
  }
}'
`;
